Feature/rabbitmq support

aslishemesh opened the pull request:
Updated unitests and open bugs.

@caduri plz review

aslishemesh committed:
update...

aslishemesh committed:
update files

aslishemesh committed:
add requirements file

aslishemesh committed:
* Update unitests (partiall)
* Fixing bugs

caduri commented on fox_scraper/fox_scraper.py:
```diff
@@ -38,6 +40,28 @@ def print_item(self):
+        self.channel = self.connection.channel()
+        self.channel.exchange_declare(exchange='fox_scrap', type='direct')
+
+#    def establish_rabbitmq_connection(self):
```
remove comments

aslishemesh replied:
will do
(when the tabbitmq module is ready - it still in "basic developing" :)  )

caduri commented on fox_scraper/fox_scraper.py:
```diff
@@ -38,6 +40,28 @@ def print_item(self):
+#        self.channel = self.connection.channel()
+#        self.channel.exchange_declare(exchange='fox_scrap', type='direct')
+
+    def close_rabbitmq_connection(self):
```
I think `close_connection` would be better name

aslishemesh replied:
agreed, I still didn't do the "change name thing" for the rabbitmq :)

caduri commented on fox_scraper/fox_scraper.py:
```diff
@@ -38,6 +40,28 @@ def print_item(self):
+    def __init__(self):
+        self.connection = pika.BlockingConnection(pika.ConnectionParameters(host='localhost'))
+        self.channel = self.connection.channel()
+        self.channel.exchange_declare(exchange='fox_scrap', type='direct')
```
Why did you chose direct?

aslishemesh replied:
I did not work on it yet.
I will choose the fanout and I will write the reasons in the code.

caduri commented on fox_scraper/fox_scraper.py:
```diff
@@ -38,6 +40,28 @@ def print_item(self):
+        self.connection.close()
+
+    def send_message_to_rabbitmq(self, message):
+        message.print_item()
```
what is that?

aslishemesh replied:
test print - told you - it is not ready.

caduri commented on fox_scraper/fox_scraper.py:
```diff
@@ -38,6 +40,28 @@ def print_item(self):
+    def close_rabbitmq_connection(self):
+        self.connection.close()
+
+    def send_message_to_rabbitmq(self, message):
```
I think `send_message` would be better name

aslishemesh replied:
ok

caduri commented on fox_scraper/fox_scraper.py:
```diff
@@ -38,6 +40,28 @@ def print_item(self):
+
+    def send_message_to_rabbitmq(self, message):
+        message.print_item()
+        body = json.dump([m.__dict__ for m in message])
```
message is a list?

caduri replied:
another thing `json.stringify` converts json to string

aslishemesh replied:
message is an FoxItem class - it still not working.
but a little question - can I use other packages? (helper pkg...)

caduri replied:
you can add a function in `FoxItem` called `to_json` and change this line to
`body = json.dumps(message.to_json())`

caduri replied:
and yes you can use whatever pkgs you feel will help you

caduri commented on fox_scraper/fox_scraper.py:
```diff
@@ -48,63 +72,54 @@ def __init__(self, site_address):
     def get_web_content(self, http_address):
```
The name suggest that you're returning the content of the site but instead its returning the scraper object

aslishemesh replied:
yes and no...
it function get the web content and returns a parser way to read this content...
so... maybe...

caduri replied:
im not sure im following, this function gets `http_address` which suggest it gets an url not the web content...

aslishemesh replied:
I changed it to get_parsed_web_contnet

caduri commented on fox_scraper/fox_scraper.py:
```diff
@@ -48,63 +72,54 @@ def __init__(self, site_address):
         :return: a list of items (Item class) of the entire catalog
         """
-        soup_main_categories = self.scrap_fox_main_categories()
+        soup_main_categories = self.get_main_categories()
```
I think `main_categories` or `fox_main_categories` would be better name

aslishemesh replied:
you right, but I used those names for more intuitive read of the code...
I will change it.

aslishemesh replied:
done

caduri commented on fox_scraper/fox_scraper.py:
```diff
@@ -48,63 +72,54 @@ def __init__(self, site_address):
-            catalog.append(self.scrap_fox_catalog(soup_sub_categories, main_cat_num))
-            main_cat_num += 1
+        for main_cat_num in range(len(soup_main_categories)): # for every main-category
+            soup_sub_categories = self.get_sub_categories(soup_main_categories[main_cat_num], main_cat_num)
```
`sub_categories`

aslishemesh replied:
ok

aslishemesh replied:
done

caduri commented on fox_scraper/fox_scraper.py:
```diff
@@ -48,63 +72,54 @@ def __init__(self, site_address):
-            main_cat_num += 1
+        for main_cat_num in range(len(soup_main_categories)): # for every main-category
+            soup_sub_categories = self.get_sub_categories(soup_main_categories[main_cat_num], main_cat_num)
+            catalog.append(self.get_full_catalog(soup_sub_categories, main_cat_num))
```
separate this into 2 lines

aslishemesh replied:
separate what?
the "main_cat_num=+1" ? why?

caduri replied:
this line `catalog.append(self.get_full_catalog(soup_sub_categories, main_cat_num))` to
```
sub_catalog = self.get_full_catalog(soup_sub_categories, main_cat_num)
catalog.append(sub_catalog)
```

aslishemesh replied:
oops :) I'm guessing you meant the append...
done.

caduri commented on fox_scraper/fox_scraper.py:
```diff
@@ -48,63 +72,54 @@ def __init__(self, site_address):
         main_categories = []
         for link in soup_main_category.find_all(class_='link padding_hf_v '):
             main_categories.append(link.get('href'))
             self.main_categories.append(link.get_text().strip())
```
why do you have a list `main_categories` and `self.main_categories`?

aslishemesh replied:
the self.main_categories is a list of categories names
the main categories is a list of categories links

caduri replied:
and why do they have the same name?

caduri commented on fox_scraper/fox_scraper.py:
```diff
@@ -48,63 +72,54 @@ def __init__(self, site_address):
         return main_categories
 
-    def scrap_fox_sub_categories(self, soup_category, category_num):
+    def get_sub_categories(self, soup_category, category_num):
```
what is the purpose of this function? to retrieve the sub_categories links? or to retrieve the catalogue for one of the main categories?

aslishemesh replied:
it retrieves the sub-categories for one of the main categories

aslishemesh replied:
I changed the names and comment

caduri commented on fox_scraper/fox_scraper.py:
```diff
@@ -48,63 +72,54 @@ def __init__(self, site_address):
+        sub_categories = self.get_web_content(str(soup_category))
         sub_catalog = []
         self.sub_categories.append(category_num)
         self.sub_categories[category_num] = []
```
you can replace this 2 lines with `self.sub_categories.append([])`, it will do the same thing.

aslishemesh replied:
ok I will

aslishemesh replied:
done, thanks.

caduri commented on fox_scraper/fox_scraper.py:
```diff
@@ -116,20 +131,34 @@ def verify_encoding(self, item_name):
             return item_name
 
-    def scrap_fox_catalog(self, soup_sub_categories, main_cat_num):
+    def get_full_catalog(self, soup_sub_categories, main_cat_num):
```
the name suggest that it will return all catalogue not just for a specific category

aslishemesh replied:
I will change it

aslishemesh replied:
done

caduri commented on fox_scraper/fox_scraper.py:
```diff
@@ -116,20 +131,34 @@ def verify_encoding(self, item_name):
+            soup_cat_sub_category = self.get_web_content(soup_sub_categories[sub_cat_num])
+            items += self.get_sub_category_catalog(soup_cat_sub_category, main_cat_num, sub_cat_num)
+        return items
+        # for soup_sub_category in soup_sub_categories:
```
remove comments

aslishemesh replied:
ok

caduri commented on tests/fox_scraper_tests.py:
```diff
@@ -1,12 +1,75 @@
+    # I created a new function in FoxItem class that returns all fields as complete string....
+    # and... it look really f..up
+    # what is the proper way to test it? I thought maybe compare the list using a simple "if" statement
+    # (it works) and then use the assert_equal on the result. but it's feels like "cheating"
```
You can implement `__eq__` in FoxItem and then iirc you can write `assert_equals` on the 2 lists

aslishemesh replied:
I will try,
but what is it iirc?

caduri replied:
iirc = if I remember correctly

aslishemesh replied:
you were right, I had a problem with my input...
the __eq__ worked fine with the __dict__
thanks.

caduri commented on .gitignore:
```diff
@@ -0,0 +1,4 @@
+*.pyc
+.idea
+.DS_Store
```
add here tests/.DS_Store as well and remove the file from git

aslishemesh replied:
ok

aslishemesh replied:
done

caduri reviewed (changes requested):
I think you need to go over the code and read it yourself, you are trying to rush things and thats not the point of this exercise.
You need to be happy with your code, and not just to make it work.

aslishemesh committed:
Updates:
	* fix general var and functions names
	* update class locations - create an helper file
	* fix the "sending" function
	* update and fix all tests

caduri commented on fox_scraper/__init__.py:
```diff
@@ -0,0 +1,3 @@
+from fox_scraper import Scraper
```
For next time, if you are under project called `fox_scraper` you don't need to add a prefix `fox_` to every file

aslishemesh replied:
ok

caduri commented on fox_scraper/fox_helper.py:
```diff
@@ -0,0 +1,65 @@
+
+    def print_item(self):
+        """
+        temp for testing - will be decrypt
```
if its temp then delete it

aslishemesh replied:
done.

caduri commented on fox_scraper/fox_helper.py:
```diff
@@ -0,0 +1,65 @@
+        convert the FoxItem into json supported type
+        :return: a dictionary to support json dumps
+        """
+        return self.__dict__
```
Don't return the `__dict__`, its not good practice.
In a case you have more attributes that some are private, `__dict__` will include them, instead create your own dict and return that.

aslishemesh replied:
ok. done

caduri commented on fox_scraper/fox_helper.py:
```diff
@@ -0,0 +1,65 @@
+        """Override the default Equals behavior to support assert-equal"""
+        if isinstance(other, self.__class__):
+            return self.__dict__ == other.__dict__
+        return NotImplemented
```
You should throw `TypeError`, I think its the best fit.

aslishemesh replied:
done

caduri commented on fox_scraper/fox_helper.py:
```diff
@@ -0,0 +1,65 @@
+    def __eq__(self, other):
+        """Override the default Equals behavior to support assert-equal"""
+        if isinstance(other, self.__class__):
+            return self.__dict__ == other.__dict__
```
It is not a good practice to use `__dict__`.
Look at the footnote here https://docs.python.org/3/tutorial/classes.html#id2
and look at this stackoverflow q&a http://stackoverflow.com/questions/4727551/accessing-python-instance-variables-with-dict-is-it-wrong

aslishemesh replied:
done

caduri commented on fox_scraper/fox_helper.py:
```diff
@@ -0,0 +1,65 @@
+        """Define a non-equality test (also - supporting assert-functions"""
+        if isinstance(other, self.__class__):
+            return not self.__eq__(other)
+        return NotImplemented
```
same comment as `__eq__`

aslishemesh replied:
ok

caduri commented on fox_scraper/fox_helper.py:
```diff
@@ -0,0 +1,65 @@
+        return NotImplemented
+
+
+class FoxSender(object):
```
No need to add a prefix

aslishemesh replied:
what do u mean?

caduri commented on fox_scraper/fox_helper.py:
```diff
@@ -0,0 +1,65 @@
+        """
+        self.connection = pika.BlockingConnection(pika.ConnectionParameters(host='localhost'))
+        self.channel = self.connection.channel()
+        self.channel.exchange_declare(exchange='fox_scrap', type='fanout')
```
A question, how can you make this utility class more generic?

aslishemesh replied:
We can use __init__ to configure the exchange name and type.
The original class had those attributes. But, since we already talked about it - we decided that we won't need any changes to the exchange.

caduri replied:
Nice :)

caduri commented on fox_scraper/fox_helper.py:
```diff
@@ -0,0 +1,65 @@
+
+    def send_message(self, message):
+        body = json.dumps(message.to_json())
+        self.channel.basic_publish(exchange='fox_scrap', routing_key='', body=body,  properties=pika.BasicProperties(delivery_mode=2))
```
what does `delivery_mode=2` do?

aslishemesh replied:
it makes sure that the messages persistent (in case the rabbitmq crashes)

caduri replied:
Didn't know that 👍

caduri commented on fox_scraper/fox_scraper.py:
```diff
@@ -14,97 +14,75 @@
     def __init__(self, site_address):
         self.site_address = site_address
         self.main_categories = []
         self.sub_categories = []
```
I would implement these 2 lists as dicts

caduri commented on fox_scraper/fox_scraper.py:
```diff
@@ -14,97 +14,75 @@
-            soup_sub_categories += list(self.scrap_fox_sub_categories(soup_main_category, main_cat_num))
-            catalog.append(self.scrap_fox_catalog(soup_sub_categories, main_cat_num))
-            main_cat_num += 1
+        for main_cat_num in range(len(main_categories_links)):
```
Next time, try to avoid passing list indexes to other functions

caduri commented on fox_scraper/fox_scraper.py:
```diff
@@ -14,97 +14,75 @@
-        for link in soup_main_category.find_all(class_='link padding_hf_v '):
-            main_categories.append(link.get('href'))
+        main_categories_content = self.get_parsed_web_content(self.site_address)
+        main_categories_links = []
```
If `self.main_categories` was a dict, you wouldn't need this list.

caduri commented on fox_scraper/fox_scraper.py:
```diff
@@ -14,97 +14,75 @@
+        sub_categories = []
+        self.sub_categories.append([])
+        for sub_category in sub_categories_content.find_all('option'):
+            current_sub_category = self.verify_item_encoding((sub_category.get_text()).strip())
```
Split this to 2 lines `(sub_category.get_text()).strip()` and dont afraid to use variables

caduri commented on fox_scraper/fox_scraper.py:
```diff
@@ -14,97 +14,75 @@
-            self.sub_categories[category_num].append(self.verify_encoding((sub_category.get_text()).strip()))
-            sub_catalog.append(format(sub_category['value']))
-        return sub_catalog
+        sub_categories_content = self.get_parsed_web_content(str(category_link))
```
why do you need to convert the link to string?

aslishemesh replied:
I don't - I used it when I though there was a different between the string and the unicode
(but it works on both)
updated.

caduri commented on fox_scraper/fox_scraper.py:
```diff
@@ -14,97 +14,75 @@
+        self.sub_categories.append([])
+        for sub_category in sub_categories_content.find_all('option'):
+            current_sub_category = self.verify_item_encoding((sub_category.get_text()).strip())
+            self.sub_categories[category_num].append(current_sub_category)
```
use dict instead list of lists

caduri commented on fox_scraper/fox_scraper.py:
```diff
@@ -14,97 +14,75 @@
+        for sub_category in sub_categories_content.find_all('option'):
+            current_sub_category = self.verify_item_encoding((sub_category.get_text()).strip())
+            self.sub_categories[category_num].append(current_sub_category)
+            sub_categories.append(format(sub_category['value']))
```
what does format do?

aslishemesh replied:
it chooses the 'valus' text from the sub-category "option" tag

aslishemesh replied:
btw - it just another way to work with strings...

caduri commented on fox_scraper/fox_scraper.py:
```diff
@@ -116,35 +94,60 @@ def verify_encoding(self, item_name):
+        """
+        items = []
+        for link in cat_sub_category.find_all(class_='box padding_hf sp_p_padding_hf center border_b border_l margin_b_1'):
+            items.append(self.parse_item(link, main_cat_num, sub_cat_num))
```
split it

caduri commented on fox_scraper/fox_scraper.py:
```diff
@@ -116,35 +94,60 @@ def verify_encoding(self, item_name):
+    # TODO - change location to new "program" function and not here
+    def send_all_items(self, catalog):
+
+        rabbit_sender = FoxSender()
```
learn on `with` and see if you can use it here

aslishemesh replied:
we can use the "with" statement to create a try-catch statement within the class (basically).
but.. we need to create __enter__() which will replace the __init__ to create the connection and __exit__() to close the connection

caduri commented on tests/fox_scraper_tests.py:
```diff
@@ -1,12 +1,81 @@
+from fox_scraper import FoxItem
+from bs4 import BeautifulSoup
+from tests.items_src import catalog
 
```
why did you use functions instead of a test class

aslishemesh replied:
because this is how the tutorial showed (https://learnpythonthehardway.org/book/ex47.html)
this is how it works with the nosetests.

caduri replied:
Next time, it won't be an excuse to why used method a and not b.
For now, http://pythontesting.net/framework/nose/nose-introduction/#example here you have an example of unitests in a class instead of functions
http://ivory.idyll.org/articles/nose-intro.html this another nice intro on nose

caduri commented on fox_scraper/fox_scraper.py:
```diff
@@ -14,97 +14,75 @@
-        print "item price: %.2f " %self.item_price
-
 
 class Scraper:
```
It would be nice if you add here a docstring with the explanation of how to use this class

caduri replied:
Nice as for next time :)

aslishemesh replied:
if u referring the Scrapper class - there is a full description couple lines up...
or i misunderstood u?

caduri replied:
You're right.

caduri commented on fox_scraper/fox_scraper.py:
```diff
@@ -14,97 +14,75 @@
+        :param category_link: sub-category link from the site
         :param category_num: category number for item description
-        :return: a list of catalog items per sub-category
+        :return: a list of sub-categories link per main-category
```
It would be nice if you had here how the return list would look like

caduri replied:
It's a general comment for all related functions

caduri replied:
And nice as for next time :)
